plugins/prometheus/prometheus.go

```diff
@@ -0,0 +1,174 @@
+
+	registerOnce.Do(func() {
+		// the server may start and stop, but the handler must only be
+		// registered once
+		http.HandleFunc("/metrics", serveMetrics)
+	})
+
+	// err is declared to avoid listener being locally scoped
+	var err error
+	listener, err = net.Listen("tcp", addr)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		err = http.Serve(listener, nil)
+	}()
+
+	return err
+}
+
+func stopPromServer(c *cli.Context) error {
+	log.Info("Halting Prometheus server.")
+	return listener.Close()
+}
+
+func promConnector(metrics []producers.MetricsMessage, c *cli.Context) error {
+	latestMetrics.Lock()
+	latestMetrics.metrics = metrics
+	latestMetrics.Unlock()
+	return nil
+}
+
+func serveMetrics(w http.ResponseWriter, r *http.Request) {
+	// operate on a copy of metrics to avoid TOCTOU issues
+	metrics := []producers.MetricsMessage{}
+	latestMetrics.Lock()
+	metrics = latestMetrics.metrics
+	latestMetrics.Unlock()
+
+	if len(metrics) == 0 {
+		http.Error(w, "", http.StatusNoContent)
+		return
+	}
+	for _, m := range metrics {
+		fmt.Fprintf(w, messageToPromText(m))
+	}
+}
+
+// messageToPromText converts a single metrics message to prometheus-formatted
+// newline-separate strings
+func messageToPromText(message producers.MetricsMessage) string {
+	var buffer bytes.Buffer
+
+	for _, d := range message.Datapoints {
+		name := strings.Replace(d.Name, ".", "_", -1)
+		labels := getLabelsForDatapoint(message.Dimensions, d.Tags)
+		t, err := time.Parse(time.RFC3339, d.Timestamp)
+		if err != nil {
+			log.Warnf("Encountered bad timestamp, %q: %s", d.Timestamp, err)
+			continue
+		}
+		timestampMs := int(t.UnixNano() / 1000000)
+		buffer.WriteString(fmt.Sprintf("%s%s %v %d\n", name, labels, d.Value, timestampMs))
+	}
+
+	return buffer.String()
+}
+
+// getLabelsForDatapoint returns prometheus-formatted labels from a
+// datapoint's dimensions and tags
+func getLabelsForDatapoint(dimensions producers.Dimensions, tags map[string]string) string {
+	allDimensions := map[string]string{}
+	if dimensions.Labels != nil {
+		allDimensions = dimensions.Labels
+	}
+
+	allDimensions["mesos_id"] = dimensions.MesosID
+	allDimensions["cluster_id"] = dimensions.ClusterID
+	allDimensions["container_id"] = dimensions.ContainerID
+	allDimensions["framework_name"] = dimensions.FrameworkName
+	allDimensions["framework_id"] = dimensions.FrameworkID
+	allDimensions["framework_role"] = dimensions.FrameworkRole
+	allDimensions["framework_principal"] = dimensions.FrameworkPrincipal
+	allDimensions["task_name"] = dimensions.TaskName
+	allDimensions["task_id"] = dimensions.TaskID
+	allDimensions["hostname"] = dimensions.Hostname
+
+	labels := []string{}
+	for k, v := range allDimensions {
+		if len(v) > 0 {
+			labels = append(labels, fmt.Sprintf("%s:%q", k, v))
+		}
+	}
+	for k, v := range tags {
+		labels = append(labels, fmt.Sprintf("%s:%q", k, v))
+	}
+
+	if len(labels) > 0 {
+		return "(" + strings.Join(labels, ",") + ")"
```

Comment:
I believe that per https://prometheus.io/docs/instrumenting/exposition_formats/#text-format-details the characters used to separate labels from the metric name are ```{}``` and not ```()```

Reply:
Thanks for catching this!